fix build on OmniOS/Solaris

rwestphal opened the pull request:
1 - OmniOS doesn't have vdprintf(), use fdopen()+vfprintf() to
    workaround this problem;
2 - Remove the _XOPEN_SOURCE definition in src/user_types/user_date_and_time.c
    to fix this compiler error: "Compiler or options invalid for pre-UNIX 03
    X/Open applications and pre-2001 POSIX applications". This change doesn't
    affect libyang build on Debian 9.3 and OpenBSD 6.3.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

michalvasko commented on src/printer.c:
```diff
@@ -84,7 +84,15 @@ ly_print(struct lyout *out, const char *format, ...)
+            count = vfprintf(stream, format, ap);
+            fclose(stream);
+        }
+#endif
```
Why the curly brackets? We try to keep case code without redundant indents so please remove them. Also, `fdopen` error checking would be nice.

rwestphal replied:
I created a new block because I needed to declare a new variable (`FILE *stream`). But I agree it's ugly, I'll add another `#ifdef` and declare the variable in the top of the function.

michalvasko replied:
Right, forgot about the variable declaration but it will probably be better to put it separately anyway.

michalvasko commented on CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@ cmake_minimum_required(VERSION 2.8.12)
 project(libyang C)
 
 include(GNUInstallDirs)
+INCLUDE(CheckSymbolExists)
```
Just a detail, please keep `INCLUDE` lowercase so that we do not mix it.

rwestphal replied:
Indeed, will update.

rwestphal commented:
Looks like my changes in `src/printer.c` caused several problems. I believe I'll need to include a new header file to have access to `HAVE_VDPRINTF`. Will take a look in a few hours...

rwestphal commented:
Done :)

Added a `dup()` inside `fdopen()` to prevent `fclose()` from closing the underlying file descriptor. I also added a `add_definitions()` line in _CMakeLists.txt_ to make the `HAVE_VDPRINTF` constant visible in the source code.

michalvasko commented:
Hi,
seems fine to me now, merging.

Regards,
Michal
